caikit/runtime/work_management/abortable_context.py

```diff
@@ -0,0 +1,178 @@
+
+    Children of this class can bind to any notification tool (e.g. grpc context)
+    """
+
+    @abc.abstractmethod
+    def must_abort(self) -> bool:
+        """Indicate whether or not the action must be aborted"""
+
+    @abc.abstractmethod
+    def set_context(self, context: AbortableContextBase):
+        """Set the abortable context that must be notified to abort work"""
+
+    @abc.abstractmethod
+    def unset_context(self):
+        """Unset any abortable context already held. Do not notify it that work should abort"""
+
+
+class ThreadInterrupter:
+    """This class implements a listener which will observe all ongoing work in `AbortableContexts`
+     and raise exceptions in the working threads if they need to be aborted.
+
+    This offers a performance advantage over the old `AbortableActions`, since only one extra
+    listener thread is created that lives for the whole lifetime of the program. The caveat
+    is that all work to be aborted must be running in a thread that is safe to kill: this may
+    not be safe to use with asyncio tasks running in an event loop.
+    """
+
+    def __init__(self):
+        # Using a SimpleQueue because we don't need the Queue's task api
+        self._queue = SimpleQueue()
+        self._thread = None
+
+        self._context_thread_map: Dict[uuid.UUID, int] = {}
+
+        self._shutdown = True
+
+    def start(self):
+        if not self._shutdown:
+            log.debug("ThreadInterrupter already started")
+            return
+        log.debug("Starting ThreadInterrupter")
+        self._shutdown = False
+        self._thread = threading.Thread(target=self._watch_loop)
+        self._thread.start()
+
+    def stop(self):
+        if self._shutdown:
+            log.debug("ThreadInterrupter already shut down")
+            return
+
+        log.info("Stopping ThreadInterrupter")
+        self._shutdown = True
+        self._queue.put(0)
+        self._thread.join(timeout=1)
+
+    def register(self, context_id: uuid, thread: int) -> None:
+        self._context_thread_map[context_id] = thread
+
+    def unregister(self, context_id: uuid) -> None:
+        self._context_thread_map.pop(context_id, None)
+
+    def kill(self, context_id: uuid) -> None:
+        # Put this context onto the queue for abortion and immediately return
+        self._queue.put(context_id, block=False)
+
+    def _watch_loop(self):
+        while True:
+            try:
+                log.debug4("Waiting on any work to abort")
+                context_id = self._queue.get()
+
+                if self._shutdown:
+                    log.debug4("Ending abort watch loop")
+                    return
+
+                self._kill_thread(context_id)
+
+                # Ensure this context/thread pair is unregistered
+                self.unregister(context_id)
+
+            except Exception:
+                log.warning("Caught exception while running abort loop", exc_info=True)
```

Comment:
Is this something we expect to happen, or is it just guarding against errors in the implementation of the loop? I would expect that the loop itself is _explicitly_ exception-free, so you wouldn't need this `try/except` and if you did happen to catch an exception, it would be fatal since it's not anticipated.

Reply:
> it would be fatal since it's not anticipated

Yeah... true. I was trying to be more defensive and not crash anything if something fails in the abort process, but I guess that if we crash the interrupter here then the server could still keep chugging along 🤷

Reply:
I think that if the interrupter crashes, we _want_ the server to go down, right? If the server is booted with the interrupter and something in our code happens to be borked and raise an exception, that's a fatal coding problem that should crash the server and trigger a bug report.

Reply:
maybe, but that could also mean very sad users!

I think you're right though that nothing in here should raise, there's really just a queue get, a dict get, and a ctypes call that should all be safe